kedro/contrib/io/pyspark/spark_hive_data_set.py

```diff
@@ -0,0 +1,306 @@
+        self._valid_write_modes = ["insert", "upsert", "overwrite"]
+        if write_mode not in self._valid_write_modes:
+            raise DataSetError(
+                "Invalid write_mode provided: {invalid}. Write_mode must be one of {valid}".format(
+                    invalid=write_mode, valid=self._valid_write_modes
+                )
+            )
+        self._write_mode = write_mode
+        if self._write_mode == "upsert" and not table_pk:
+            raise DataSetError("table_pk must be set to utilise upsert read mode")
+        self._table_pk = table_pk
+
+        self._table_columns = self._load().columns if self._exists() else None
+
+        if (
+            self._table_pk
+            and self._exists()
+            and not all(
+                (col_name in self._table_columns for col_name in self._table_pk)
+            )
+        ):
+            raise DataSetError(
+                "columns [{colnames}] selected as PK not found in table {database}.{table}".format(
+                    colnames=", ".join(self._table_pk),
+                    database=self._database,
+                    table=self._table,
+                )
+            )
+
+    @staticmethod
+    def _get_spark():
+        return SparkSession.builder.getOrCreate()
+
+    def _create_empty_hive_table(self, data):
+        data.createOrReplaceTempView("tmp")
+        self._get_spark().sql(
+            "create table {database}.{table} select * from tmp limit 1".format(
+                table=self._table, database=self._database
+            )
+        )
+        self._get_spark().sql(
+            "truncate table {database}.{table}".format(
+                database=self._database, table=self._table
+            )
+        )
+
+    def _load(self) -> DataFrame:
+        if not self._exists():
+            raise DataSetError(
+                "requested table not found: {database}.{table}".format(
+                    database=self._database, table=self._table
+                )
+            )
+        return self._get_spark().sql(
+            "select * from {database}.{table}".format(
+                database=self._database, table=self._table
+            )
+        )
+
+    def _save(self, data: DataFrame) -> None:
+        if not self._exists():
+            self._create_empty_hive_table(data)
+            self._table_columns = data.columns
+        self._validate_save(data)
+        write_methods = {
+            "insert": self._insert_save,
+            "upsert": self._upsert_save,
+            "overwrite": self._overwrite_save,
+        }
+        write_methods[self._write_mode](data)
+
+    def _insert_save(self, data):
+        data.createOrReplaceTempView("tmp")
+        self._get_spark().sql(
+            "insert into {database}.{table} select {columns} from tmp".format(
+                database=self._database,
+                table=self._table,
+                columns=", ".join(self._table_columns),
```

Comment:
Not sure whether this is in scope of this PR, but it might be quite vulnerable to SQL injections (depending on where the parameters for dataset instantiation come from)

Reply:
This variable is built by reading the columns of the dataframe provided using `data_frame.columns` to be written or the one on the hive DB. I feel reasonably comfortable assuming those should be safe to inject into this code but could investigate other means if needed.